kratos/utilities/normal_calculation_utils.cpp

```diff
@@ -476,25 +481,47 @@ void NormalCalculationUtils::CalculateNormalsUsingGenericAlgorithm(
     auto& r_entities_array = GetContainer<TContainerType>(rModelPart);
     const auto it_entity_begin = r_entities_array.begin();
 
-    // TODO: Use TLS in ParallelUtilities
-#pragma omp parallel for firstprivate(aux_coords)
-    for (int i = 0; i < static_cast<int>(r_entities_array.size()); ++i) {
-        auto it_entity = it_entity_begin + i;
-        auto& r_geometry = it_entity->GetGeometry();
+    if (ConsiderUnitNormal) {
+        // TODO: Use TLS in ParallelUtilities
+        #pragma omp parallel for firstprivate(aux_coords)
+        for (int i = 0; i < static_cast<int>(r_entities_array.size()); ++i) {
+            auto it_entity = it_entity_begin + i;
+            auto& r_geometry = it_entity->GetGeometry();
 
-        // Avoid not "flat" elements
-        if (r_geometry.WorkingSpaceDimension() != r_geometry.LocalSpaceDimension() + 1) {
-            continue;
+            // Avoid not "flat" elements
+            if (r_geometry.WorkingSpaceDimension() != r_geometry.LocalSpaceDimension() + 1) {
+                continue;
+            }
+
+            // Iterate over nodes
+            for (auto& r_node : r_geometry) {
+                r_geometry.PointLocalCoordinates(aux_coords, r_node.Coordinates());
+                r_node.SetLock();
+                noalias(r_node.FastGetSolutionStepValue(NORMAL)) += r_geometry.UnitNormal(aux_coords);
+                r_node.UnSetLock();
+            }
         }
+    } else {
+        // TODO: Use TLS in ParallelUtilities
+        #pragma omp parallel for firstprivate(aux_coords)
+        for (int i = 0; i < static_cast<int>(r_entities_array.size()); ++i) {
+            auto it_entity = it_entity_begin + i;
+            auto& r_geometry = it_entity->GetGeometry();
+
+            // Avoid not "flat" elements
+            if (r_geometry.WorkingSpaceDimension() != r_geometry.LocalSpaceDimension() + 1) {
+                continue;
+            }
 
-        // Iterate over nodes
-        const double coefficient = 1.0 / static_cast<double>(r_geometry.PointsNumber());
-        for (auto& r_node : r_geometry) {
-            r_geometry.PointLocalCoordinates(aux_coords, r_node.Coordinates());
-            const auto normal = r_geometry.Normal(aux_coords);
-            r_node.SetLock();
-            noalias(r_node.FastGetSolutionStepValue(NORMAL)) += normal * coefficient;
-            r_node.UnSetLock();
+            // Iterate over nodes
+            const double coefficient = 1.0 / static_cast<double>(r_geometry.PointsNumber());
+            for (auto& r_node : r_geometry) {
+                r_geometry.PointLocalCoordinates(aux_coords, r_node.Coordinates());
+                const auto normal = r_geometry.Normal(aux_coords);
```

Comment:
this line is the only difference in the loops right?
=> maybe you can create a lambda to get the corresponding normal instead copy pasting the entire loop

Reply:
Ye, I was thinking about that, but it would be slower?

Reply:
Let me check

Reply:
hm not sure
I would hope not but of this might be only a wish...
@roigcarlo any idea?

Reply:
Done